### ISSUE TITLE
Private K8s cluster on Azure

### ISSUE THREAD

**Issue by monojit18** [kubernetes-sigs/cluster-api-provider-azure#1495]
As per CAPZ docs, Management cluster should be in a peered VNET or same VNET as workload cluster.

But since we are creating the Management cluster thru _Kind_ as a local cluster then how to achieve this?

Can the Management cluster be another light-weight AKS cluster may be which would run in the same VNET as Workload cluster, to achieve this?

Right now Private cluster creation is failing as expected!

**Comment by devigned** [comment]
The end to end test suite of CAPZ contains a private cluster test.
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/300757bdc09eee1ff5be83eb1684c55fb15be394/test/e2e/azure_privatecluster.go#L56

In that test, the initial management cluster is run in kind, which creates a publicly exposed cluster in Azure, then pivots the management cluster into the newly created Azure workload cluster. From there, the new Azure management cluster is used to create a private workload cluster with shared connectivity within the VNet.

If one were to build a totally private management and workload clusters (both clusters with no access from outside of their VNet, no inbound access from the public network), one would need to execute the kind bootstrapping from a machine within the Azure VNet or with VPN access established to the Azure VNet. One way or another, the kind bootstrap cluster needs to have access to the first cluster created in Azure. We currently don't have a strong opinion about how this access is achieved. 

We are interested in improving the private bootstrapping experience. If you have ideas about how to improve the experience, please share them.

**Comment by CecileRobertMichon** [comment]
> Can the Management cluster be another light-weight AKS cluster may be which would run in the same VNET as Workload cluster, to achieve this?

yes

Basically, you need some other k8s cluster that is in the same Azure VNET or a peered VNET (and thus has access to the private API endpoint of the private cluster). That cluster can be an AKS cluster, an AKS Engine cluster, another CAPZ cluster, or even a kind or Minikube cluster running on an Azure VM. I also know of someone who build a cloud init script to install CAPZ on an Azure VM using MicroK8s, you could use it as inspiration: https://github.com/scotty-c/cluster-api-azure-control-plane

**Comment by monojit18** [comment]
This is now blocked due to the issue: #1494 

This is what I have tried:

Created an AKS cluster with 2 Nodes, 1 VNET and 1 Subnet - Management Cluster for CAPZ
Created 1 VNET for CAPZ - _18.0.0.0/23_

Created 2 Subnets (Master, Worker) - _18.0.0.0/24, 18.0.1.0/24_ 
Peered AKS VNET and CAPZ VNET

Tried Creating cluster with CAPZ templates using 
_apiServerLB:
      type: Internal
      frontendIPs:
      - name: lb-private-ip-frontend
        privateIP: 18.0.0.100_

The cluster creation is going on forever as described in #1494.

So, basically my private cluster creation fails with this approach
Can you please help/guide me here?
Anything I am missing here?

**Comment by monojit18** [comment]
Team,
Could not get this working as Private Cluster :(

As #1494 was resolved by adding an NSG to control plane subnet; I have added one additional step:
Created a Private DNS Zone - <cluster-name>.capz.io and linked the Management Cluster VNet into the list of Virtual Network links

But now the Private cluster creation is stuck at the Provisioning stage of the Control Plane VM
Attached log for reference.
[log.txt](https://github.com/kubernetes-sigs/cluster-api-provider-azure/files/7123179/log.txt)

Can you please let me know what step is missing for the private cluster creation?

**Comment by CecileRobertMichon** [comment]
@monojit18 if you are using a peered vnet, did you also link the peered vnet to the private DNS zone? I ask because this issue seems very similar to the one @sshiba ran into recently: https://kubernetes.slack.com/archives/CEX9HENG7/p1628524704080700 

You could check that this is done properly by running `nslookup  apiserver.target-cluster.capz.io` from the management cluster's vnet and making sure that it resolves to the IP of the internal API Server load balancer of the cluster.

We have https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1601 open to make this automatic but for now it needs to be done out of band in order to use a peered vnet.

**Comment by monojit18** [comment]
@CecileRobertMichon  Yes, I did that specifically. So following are my steps (_different from Public Cluster scenario_):

- Added NSG to CAPZ Vnet - _Allow SSH(22), and Allow API Server(6443)_
- Created Private DNS Zone as <cluster_name>.capz.io
- Peered Management Vnet (AKS) with the CAPZ Vnet
- Added Management Vnet to the Virtual Network links of the Private DNS Zone
- Created an Nginx POD in the Management cluster (AKS)
- _k exec -it <pod_name> -- sh_
- _nslookup apiserver.capz-k8s-cluster.capz.io_
_Server:         10.0.0.10
Address:        10.0.0.10:53_

_Name:   apiserver.capz-k8s-cluster.capz.io
Address: 16.0.0.100_

- Modified the Manifest file as:
_apiServerLB:
      type: Internal
      frontendIPs:
      - name: lb-private-ip-frontend
        privateIP: 16.0.0.100_

- Cluster creation process goes into a pending state for ever (PFA)
- Attached log.txt also which shows an error - _Bootstrap data secret reference is not yet available_

[log.txt](https://github.com/kubernetes-sigs/cluster-api-provider-azure/files/7125951/log.txt)
<img width="554" alt="Screenshot 2021-09-08 at 10 24 59 AM" src="https://user-images.githubusercontent.com/2395966/132448686-36735c65-8d75-42f9-aa49-6c9e9596cc53.png">

<img width="1385" alt="Screenshot 2021-09-08 at 9 52 41 AM" src="https://user-images.githubusercontent.com/2395966/132448318-2d0d117c-4bbb-40cb-814c-074223ae048e.png">
<img width="1393" alt="Screenshot 2021-09-08 at 9 52 55 AM" src="https://user-images.githubusercontent.com/2395966/132448327-0f4b60f0-699c-4555-9d8d-549eb887635a.png">

- Just fYI, the same steps work well with Public Cluster creation

**Comment by monojit18** [comment]
Team,

Is there anything I am still missing for Private clusters?

**Comment by CecileRobertMichon** [comment]
@monojit18 could you try with [this script](https://github.com/kubernetes-sigs/cluster-api-provider-azure/files/6979058/create-peering-vnet.txt) provided by @sshiba in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1607 ? 

@Jont828 is working on #532 and #1601 to make this whole process easier BTW, hoping that will help.

**Comment by monojit18** [comment]
@CecileRobertMichon Somehow the script did not work for me!
But I would wait for the #532 and/or #1601 to happen and check?
